core/store/src/trie/shard_tries.rs

```diff
@@ -778,13 +778,14 @@ impl WrappedTrieChanges {
                 "NotWritableToDisk changes must never be finalized."
             );
 
-            assert!(
-                !change_with_trie_key.changes.iter().any(|RawStateChange { cause, .. }| matches!(
-                    cause,
-                    StateChangeCause::Resharding
-                )),
-                "Resharding changes must never be finalized."
-            );
+            // Resharding changes must not be finalized, however they may be introduced here when we are
+            // evaluating changes for split state in process_split_state function
+            change_with_trie_key
```

Comment:
This is the place where we were getting the duplicate keys in the DB transaction. We need to filter out (not commit them to DB) the state_changes for child shards during resharding as they are already taken care by the parent shard.

But at the same time, we can not not include the state_changes field as that is/will be used by flat storage to update state.

Reply:
Can you tell me how can we end up with both Resharding and non-Resharding state changes within one WrappedTrieChanges? Does it happen when IsCaughtUp, CatchingUp or NotCaughtUp?

Reply:
Right, this was slightly difficult to debug but here's how it all happens. This is specifically a part of `IsCaughtUp` where we are processing a block and then applying the state changes to both the parent and the child blocks. Specifically, look at `process_apply_chunk_result` -> case `ApplyChunkResult::SameHeight`.

The way it works is that we call `self.chain_store_update.save_trie_changes(apply_result.trie_changes)` that has the state changes for the parent block. But in the `apply_state_changes_to_split_states`, the way it's constructed is that it breaks the state changes across the child shards as well. The state changes from the child shards get applied in the `self.process_split_state` function -> `self.chain_store_update.save_trie_changes(result.trie_changes)`

The way the state changes are represented in DB, there are not shard details for them. Which means each state change shows up twice during this processing, once from parent shard, once from child shard.

Also, this is only visible in the `IsCaughtUp` mode as we bundle the transactions from child and parent in the same transaction.